packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,341 @@
+      });
+
+      this.transport.send(jsonrpcRequest).catch(error => {
+        cleanup();
+        reject(error);
+      });
+    });
+  }
+
+  private async listTools({
+    params,
+    options,
+  }: {
+    params?: ListToolsRequest['params'];
+    options?: RequestOptions;
+  } = {}): Promise<ListToolsResult> {
+    return this.request({
+      request: { method: 'tools/list', params },
+      resultSchema: ListToolsResultSchema,
+      options,
+    });
+  }
+
+  private async callTool({
+    params,
+    resultSchema,
+    options,
+  }: {
+    params: CallToolRequest['params'];
+    resultSchema: typeof CallToolResultSchema;
+    options?: RequestOptions;
+  }): Promise<CallToolResult> {
+    return this.request({
+      request: { method: 'tools/call', params },
+      resultSchema,
+      options,
+    });
+  }
+
+  private async notification(notification: Notification): Promise<void> {
+    const jsonrpcNotification: JSONRPCNotification = {
+      ...notification,
+      jsonrpc: '2.0',
+    };
+
+    await this.transport.send(jsonrpcNotification);
+  }
+
+  /**
+   * Returns a set of AI SDK tools from the MCP server
+   * @returns A record of tool names to their implementations
+   */
+  async tools<TOOL_SCHEMAS extends ToolSchemas = 'automatic'>({
+    schemas = 'automatic',
+  }: {
+    schemas?: TOOL_SCHEMAS;
+  } = {}): Promise<McpToolSet<TOOL_SCHEMAS>> {
+    const tools: Record<string, Tool> = {};
+    const listToolsResult = await this.listTools();
+
+    for (const { name, description, inputSchema } of listToolsResult.tools) {
+      if (schemas !== 'automatic' && !(name in schemas)) {
+        continue;
+      }
+
+      const parameters =
+        schemas === 'automatic'
+          ? jsonSchema(inputSchema as JSONSchema7)
+          : schemas[name].parameters;
+
+      const toolWithExecute = tool({
+        description,
+        parameters,
+        execute: async (
+          args: inferParameters<typeof parameters>,
+          options: ToolExecutionOptions,
+        ): Promise<CallToolResult> => {
+          options?.abortSignal?.throwIfAborted();
+
+          const result = await this.callTool({
+            params: {
+              name,
+              arguments: args,
+            },
+            resultSchema: CallToolResultSchema,
+            options: {
+              signal: options.abortSignal,
+            },
+          });
+
+          return result;
+        },
+      });
+
+      tools[name] = toolWithExecute;
+    }
+
+    return tools as McpToolSet<TOOL_SCHEMAS>;
+  }
+
+  private onclose(): void {
+    const responseHandlers = this.responseHandlers;
+    this.responseHandlers = new Map();
+
+    const error = new MCPClientError({
+      message: 'Connection closed',
+    });
+
+    for (const handler of responseHandlers.values()) {
+      handler(error);
+    }
+  }
+
+  private onerror(error: Error): void {
+    throw new MCPClientError({
+      message: error.message,
+      cause: error,
+    });
+  }
```

Comment:
what will this do in practice (in terms of the stack and where the error shows up)?

Reply:
ah, good point! since the client's `onError` (which is triggered via the transport's `onerror` handler or `onResponse`) we aren't properly propagating this up within the client

Reply:
instead of throwing `MCPClientError`, should we just close the client+transport? we need a way to properly bubble up the error to the user hmmmmm